torch/_higher_order_ops/scan.py

```diff
@@ -105,72 +106,80 @@ def add(x: torch.Tensor, y: torch.Tensor):
+        
+        # Check flat_xs and flat_init for type
+        validate_subgraph_args_types(flat_xs)
+        validate_subgraph_args_types(flat_init)
+        
+        if len(flat_init) == 0:
+            raise RuntimeError("Init tensors must be provided")
+        
+        if any(x.ndim < dim for x in flat_xs):
             raise RuntimeError(
-                f"All xs leaves must have a scan dimension > 0 but got {x}"
+                "All elements of xs must at least have 'dim' number of dimensions"
             )
-
-    if len(leaves_xs) == 0:
-        return pytree.tree_unflatten(leaves_init, spec_init), xs
-
-    shape = leaves_xs[0].shape
+        if any(x.shape[dim] == 0 for x in flat_xs):
+            raise RuntimeError(
+                "The scan dimension of all elements of xs must be > 0"
+            )
+                
+        if len(flat_xs) == 0:
+            return pytree.tree_unflatten(flat_init, spec_init), xs
+        
+    _validate_input(combine_fn, flat_xs, flat_init, dim)
+
+    # def _check_new_carry_match_init(leaves_init, carry_leaves):
+    #     for i, (init, new_carry) in enumerate(zip(leaves_init, carry_leaves)):
+    #         if init.shape != new_carry.shape:
+    #             raise RuntimeError(
+    #                 f"The shape of the new_carry[{i}] {new_carry.shape} doesn't match that of the init[{i}] {init.shape}."
+    #             )
+    #         if init.stride() != new_carry.stride():
+    #             raise RuntimeError(
+    #                 f"The stride of the new_carry[{i}] {new_carry.stride()} doesn't match that of the init[{i}] {init.stride()}."
+    #             )
+    #         if init.dtype != new_carry.dtype:
+    #             raise RuntimeError(
+    #                 f"The dtype of the new_carry[{i}] {new_carry.dtype} doesn't match that of the init[{i}] {init.dtype}."
+    #             )
+    #         if init.requires_grad != new_carry.requires_grad:
+    #             raise RuntimeError(
+    #                 f"The requires_grad of the new_carry[{i}] {new_carry.requires_grad} doesn't match that of the init[{i}] {init.requires_grad}."  # noqa: B950
+    #             )
+                
+    shape = flat_xs[0].shape
     ndim = len(shape)
     dim = utils.canonicalize_dim(ndim, dim)
 
-    out = combine_fn(
-        pytree.tree_unflatten(leaves_init, spec_init),
-        pytree.tree_unflatten([elem.select(dim, 0) for elem in leaves_xs], spec_xs),
-    )
-
-    # The first output needs to have the same pytree as init
-    carry_leaves = pytree.tree_leaves(out[0])
-    if len(carry_leaves) != len(leaves_init):
-        raise RuntimeError(
-            f"The number of leaves of the pytree of the new carry produced by the operator is {len(carry_leaves)}\
-doesn't match the length of the pytree of the init {len(leaves_init)}"
-        )
+    outs = combine_fn(pytree.tree_unflatten(flat_init, spec_init),
```

Comment:
So the plan is, for the user-api Scan:
```
all_flat_args, all_inp_spec = pytree.tree_flatten((init, xs))
flat_init, init_spec = pytree.tree_flatten(init)
flat_xs, xs_spec = pytree.tree_flatten(xs)
# some input checking...
def new_combine_fn(*flat_args):
  init, xs = pytree.tree_unflatten(*flat_args, inp_spec)
  return combine_fn(init, xs)
# deal with dim and reverse...
def _scan_op_wrapper(*args, **kwargs):
  return scan_op(*args, **kwargs)
torch.compile(_scan_op_wrapper)(new_combine_fn, flat_init, flat_xs, tuple())
```
In dynamo, we call speculate_subgraph(..., should_flatten_outputs=True),
so it returns body_r (flattened) and a body_treespec (that can be used to unflatten the flat inputs).
After we get the body_r, we could check body_r and sliced init's meta match. An example to follow is a diff_meta as used here: https://github.com/pytorch/pytorch/pull/142162

Reply:
I pushed some updates which should include this. I hope it goes into the right direction.